### ISSUE TITLE
Raw map with custom ID?

### ISSUE THREAD

**Issue by axelkar** [salsa-rs/salsa#539]
Hey! Rust-analyzer has a map from `FileId` to `Arc<str>` implemented by Salsa. Is there a way to make something similar with "main salsa"? Do I have to make a `HashMap<FileId, SourceFileText>` into my database struct? How would tracking interact with it? Deleting and updating?

Edit: Are query groups upcoming features from the rust-analyzer fork or outdated features?

**Comment by MichaReiser** [comment]
I don't fully understand your use case because I'm not that familiar with rust analyzer. But I think you have two possibilities:

1. Make `File` a salsa input with a field `source_text`.
2. Make `File` a salsa input that only stores metadata about the file (last modified, path, etc), and create a `source_text(File)` query

For both cases, you'll want to keep some data structure that allows you to lookup a `File`, most likely by path. 

See https://salsa-rs.github.io/salsa/tutorial/ir.html#input-structs

**Comment by davidbarsky** [comment]
> Hey! Rust-analyzer has a map from `FileId` to `Arc<str>` implemented by Salsa

Just to confirm, are you talking about `FileLoader` and `SourceDatabase` [in `base-db`](https://github.com/rust-lang/rust-analyzer/blob/eb5da56d839ae0a9e9f50774fa3eb78eb0964550/crates/base-db/src/lib.rs#L49-L81)? If so, what Micha said, but you might also benefit looking at his workspace implementation [in ruff](https://github.com/astral-sh/ruff/tree/f37b39d6cc118203fb16d2a4d1f2bef1134519e4/crates/red_knot/src). I've certainly found it useful and I expect I'll pilfer parts of it to replace rust-analyzer's VFS! 

> Edit: Are query groups upcoming features from the rust-analyzer fork or outdated features? 

I don't know if query groups are strictly necessary in the new Salsa (Salsa 3.0), as they seem to fall out naturally from the [existing design](https://github.com/salsa-rs/salsa/pull/538#discussion_r1694250775), but don't quote me on that!

**Comment by Skepfyr** [comment]
I guess I'll add https://salsa-rs.github.io/salsa/common_patterns/on_demand_inputs.html as a potentially useful bit of documentation, depending on exactly what you're trying o achieve.

**Comment by axelkar** [comment]
>     1. Make `File` a salsa input with a field `source_text`.

So I should make a `DashMap<FileId, File>` like in the on-demand inputs example?

I had an old version of r-a cloned and got a bit confused but now I understand that every query that has FileId as an input has to query the HashMap!

>     2. Make `File` a salsa input that only stores metadata about the file (last modified, path, etc), and create a `source_text(File)` query

How does this work? I would still have to have a FileId -> entity ID map, right?

> .. and you can set the value of the field with `source.set_text(&mut db, "print 11 * 2".to_string())`.
> <https://salsa-rs.github.io/salsa/tutorial/ir.html#input-structs>

I just noticed this! Is there a way to delete entities? If I were to just create new entities for new versions of files, would the old version's entities ever get garbage-collected?

Also, since `set_text` requires a mutable reference, can I RwLock or make an RCU for my database? I'm developing a language server using `tower_lsp` and want to extend the db infra for a compiler eventually.

**Comment by MichaReiser** [comment]
> So I should make a DashMap<FileId, File> like in the on-demand inputs example?

Hard to say without understanding what `FileId` is. `File` is already a `u32`, so I'm not sure if it's worth keeping around another ID, but maybe it is. I don't know without understanding your design in more detail.

> I just noticed this! Is there a way to delete entities? If I were to just create new entities for new versions of files, would the old version's entities ever get garbage-collected?

There's currently no way to delete inputs. Ruff uses a `status: Deleted` to mark deleted files as *gone* 

*  If I were to just create new entities for new versions of files, would the old version's entities ever get garbage-collected?
No, they wouldn't and you would also loose out on caching. You need to ensure that you keep using the same inputs. 

You may want to take a look at Ruff's [Files](https://github.com/astral-sh/ruff/blob/87418a061ff49e4defadd953a316b713b62cd046/crates/ruff_db/src/files.rs#L43) or the lazy input example from  or the [on-demand input example](https://salsa-rs.github.io/salsa/common_patterns/on_demand_inputs.html) as mentioned before.

**Comment by axelkar** [comment]
Thanks to everyone!

Is there a way to snapshot and use the database in parallel on master or should I go back to commit a1bf3a613f451af7fc0a59411c56abc47fe8e8e1? What about field mutability?

**Comment by MichaReiser** [comment]
> Is there a way to snapshot and use the database in parallel on master or should I go back to commit https://github.com/salsa-rs/salsa/commit/a1bf3a613f451af7fc0a59411c56abc47fe8e8e1?


You can wrap your Db in a `Handle`

https://github.com/salsa-rs/salsa/blob/master/src/handle.rs
https://github.com/astral-sh/ruff/blob/2f54d05d975ccf1757ce4cf796eff8f635790671/crates/red_knot/src/main.rs#L127-L132

This API is most likely to change. 

> Also, since set_text requires a mutable reference, can I RwLock or make an RCU for my database? I

You can call `Handle.get_mut()` to get a mutable database (only possible outside of queries). Be aware, that calling `get_mut` cancels any pending query

**Comment by MichaReiser** [comment]
I'll close this issue. Feel free to open a new issue or reopen this one when you run into more questions

**Comment by axelkar** [comment]
@MichaReiser Do you have notifications on for this issue? I can't reopen this either.

**Comment by MichaReiser** [comment]
I have no experience integrating salsa into tower-lsp. If you want exclusive access to the database, you'll probably have to wrap `self.database` in a mutex. 

> Do you have notifications on for this issue? I

I do have notifications enabled but I'm not triaging salsa issues 24h a day ...
